File: sort_tech/merge_sort.c
#include <stdio.h>

// Function to merge two halves of the array
void merge(int arr[], int left, int mid, int right, int* step) {
    int n1 = mid - left + 1;
    int n2 = right - mid;

    // Temporary arrays to hold the divided parts
    int L[n1], R[n2];

    // Copy data to temporary arrays L[] and R[]
    for (int i = 0; i < n1; i++)
        L[i] = arr[left + i];
    for (int i = 0; i < n2; i++)
        R[i] = arr[mid + 1 + i];

    // Merge the temp arrays back into arr[left..right]
    int i = 0; // Initial index of first subarray
    int j = 0; // Initial index of second subarray
    int k = left; // Initial index of merged subarray

    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) {
            arr[k] = L[i];
            i++;
        } else {
            arr[k] = R[j];
            j++;
        }
        k++;
    }

    // Copy the remaining elements of L[], if any
    while (i < n1) {
        arr[k] = L[i];
        i++;
        k++;
    }

    // Copy the remaining elements of R[], if any
    while (j < n2) {
        arr[k] = R[j];
        j++;
        k++;
    }

    // Print the array state after merging
    printf("Step %d: ", (*step)++);
    for (int l = left; l <= right; l++) {
        printf("%d ", arr[l]);
    }
    printf("\n");
}

// Function to implement Merge Sort
void mergeSort(int arr[], int left, int right, int* step) {
    if (left < right) {
        int mid = left + (right - left) / 2;

        // Recursively divide the array
        mergeSort(arr, left, mid, step);
        mergeSort(arr, mid + 1, right, step);

        // Merge the divided parts
        merge(arr, left, mid, right, step);
    }
}

int main() {
    int n;

    // Asking the user for the number of elements
    printf("Enter the number of elements: ");
    scanf("%d", &n);

    int arr[n];

    // Taking input from the user for the array
    printf("Enter the elements: ");
    for (int i = 0; i < n; i++) {
        scanf("%d", &arr[i]);
    }

    int step = 1; // Step counter to track comparisons
    mergeSort(arr, 0, n - 1, &step);

    // Final sorted array
    printf("\nSorted array: ");
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");

    return 0;
}

File: PYTHON/merge_sort.py
# Python program for implementation of Merge Sort

def mergeSort(arr):
    # Check if the array has more than one element
    if len(arr) > 1:

        # Finding the mid index of the array
        mid = len(arr) // 2

        # Dividing the array elements into two halves
        L = arr[:mid]  # Left half
        R = arr[mid:]  # Right half

        # Recursively sorting the first half
        mergeSort(L)

        # Recursively sorting the second half
        mergeSort(R)

        # Initializing indices for left (L), right (R), and merged array (arr)
        i = j = k = 0

        # Merging the temp arrays back into arr
        while i < len(L) and j < len(R):
            # Compare elements from both halves and merge them in sorted order
            if L[i] <= R[j]:
                arr[k] = L[i]  # If left element is smaller or equal
                i += 1  # Move to the next element in left half
            else:
                arr[k] = R[j]  # If right element is smaller
                j += 1  # Move to the next element in right half
            k += 1  # Move to the next position in merged array

        # Checking if any element was left in L
        while i < len(L):
            arr[k] = L[i]  # Copy remaining elements from L
            i += 1  # Move to next element in left half
            k += 1  # Move to next position in merged array

        # Checking if any element was left in R
        while j < len(R):
            arr[k] = R[j]  # Copy remaining elements from R
            j += 1  # Move to next element in right half
            k += 1  # Move to next position in merged array

# Function to print the elements of the list
def printList(arr):
    for i in range(len(arr)):
        print(arr[i], end=" ")  # Print each element followed by a space
    print()  # Print newline after the list

# Driver Code
if __name__ == '__main__':
    arr = [12, 11, 13, 5, 6, 7]  # Sample input array
    print("Given array is")
    printList(arr)  # Print the original array
    mergeSort(arr)  # Sort the array using merge sort
    print("\nSorted array is ")
    printList(arr)  # Print the sorted array
